### PR TITLE
Closure

### PR DESCRIPTION
**Pull request by ranjitjhala**
Basic (and unsound) support for closure, mostly just makes sure flux doesn't crash by adding the 
various data types to `lower` and `conv` closures. 

**The closure body itself is completely ignored!** 

Next:

1. Generate `c_body` = Constraint for checking the closure body against its kvar-signature
2. Package (conjoin) `c_body` with the constraints at the call-site

### CODE REVIEW

File: flux-refineck/src/constraint_gen.rs [hunk 1]
```diff
@@ -388,6 +402,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 rcx.check_pred(p2, self.tag);
                 self.subtyping(rcx, ty1, ty2);
             }
+            (TyKind::Closure(did1), TyKind::Closure(did2)) if did1 == did2 => {}
```
Comment (nilehmann) [inline flux-refineck/src/constraint_gen.rs:405]: hmmm this is interesting maybe we can delay the implementation of function subtyping and it's just a matter of figuring out the `ClosureSubsts` deal.

**Review by nilehmann — COMMENTED** [review]
There's not much going on yet, but this looks ok to get the first pieces in place. The only major comment is to move `TyKind::Closure` to `BaseTy::Closure`

File: flux-middle/src/rty/mod.rs [hunk 2]
```diff
@@ -166,6 +166,7 @@ pub enum TyKind {
     /// [`Rvalue::Discriminant`]: crate::rustc::mir::Rvalue::Discriminant
     /// [`TerminatorKind::SwitchInt`]: crate::rustc::mir::TerminatorKind::SwitchInt
     Discr(AdtDef, Place),
+    Closure(DefId),
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-middle/src/rty/mod.rs:169]: done

File: flux-refineck/src/constraint_gen.rs (hunk 1 above)
Comment (ranjitjhala) [reply flux-refineck/src/constraint_gen.rs:405]: hmmm . removing for now!

File: flux-middle/src/rustc/mir.rs [hunk 3]
```diff
@@ -51,6 +51,11 @@ pub struct CallSubsts<'tcx> {
     pub lowered: List<GenericArg>,
 }
 
+pub struct ClosureSubsts<'tcx> {
+    pub orig: SubstsRef<'tcx>,
+    pub lowered: List<GenericArg>,
+}
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-middle/src/rustc/mir.rs:57]: hmm remove for now!

**Comment by nilehmann** [comment]
@ranjitjhala do you want to keep working on top of this PR or do you want to merge? I'm fine either way

**Comment by ranjitjhala** [comment]
I wanted to merge it in first as you suggested and then keep working to
actually check the closure body (in subsequent PR)

**Review by nilehmann — APPROVED** [review]
looks good then

**Pull request merged**